newton.go

```diff
@@ -160,415 +73,93 @@ func (n *Newton) sign(req *http.Request) error {
-		if err != nil {
-			log.Printf("error:%s", err.Error())
-		}
-	}()
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
-	}
-
-	body, _ := ioutil.ReadAll(res.Body)
-
-	var resp GetMinTradeAmountsResp
-	err = json.Unmarshal(body, &resp.TradeAmounts)
-	if err != nil {
-		return nil, err
-	}
-
-	return &resp, nil
-}
-
-// Private API
-///////////////////////////////////////////////////////////////////////////////////////////////////
-func (n *Newton) doPrivateQuery(path string, method string, args []Args, body string) (*http.Response, error) {
-	url := baseUrl + path
-
-	req, _ := http.NewRequest(method, url, bytes.NewBuffer([]byte(body)))
+	req, _ := http.NewRequest(
+		query.GetMethod(),
+		n.BaseUrl + query.GetPath(), 
+		bytes.NewBuffer(body))
+	
 	q := req.URL.Query()
-	for _, a := range args {
+	for _, a := range query.GetParameters() {
 		q.Add(a.Key, a.Value)
 	}
+
 	req.URL.RawQuery = q.Encode()
-	if method != http.MethodGet {
+	if query.GetMethod() != http.MethodGet {
 		req.Header.Add("content-type", "application/json")
 	}
-	err := n.sign(req)
-	if err != nil {
-		return nil, err
-	}
-	res, err := http.DefaultClient.Do(req)
-
-	return res, err
-}
-
-func (n *Newton) Balances(asset string) (*BalancesResp, error) {
-
-	a := make([]Args, 1)
 
-	a[0].Key = "asset"
-	a[0].Value = asset
-	res, err := n.doPrivateQuery("/balances", http.MethodGet, a, "")
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		err := res.Body.Close()
+	if !query.IsPublic() {
+		err := n.sign(req)
 		if err != nil {
-			log.Printf("error:%s", err.Error())
+			return nil, err
 		}
-	}()
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
 	}
 
-	body, _ := ioutil.ReadAll(res.Body)
-
-	var b BalancesResp
-	err = json.Unmarshal(body, &b.Balances)
-	if err != nil {
-		return nil, err
-	}
-
-	return &b, nil
-}
-
-func (n *Newton) Actions(actionType ActionType, limit int, offset int, startDate int64, endDate int64) (*ActionsResp, error) {
-
-	a := make([]Args, 5)
-
-	a[0].Key = "action_type"
-	a[0].Value = string(actionType)
-
-	a[1].Key = "end_date"
-	a[1].Value = strconv.FormatInt(endDate, 10)
-
-	a[2].Key = "limit"
-	a[2].Value = strconv.Itoa(limit)
-
-	a[3].Key = "offset"
-	a[3].Value = strconv.Itoa(offset)
-
-	a[4].Key = "start_date"
-	a[4].Value = strconv.FormatInt(startDate, 10)
-
-	res, err := n.doPrivateQuery("/actions", http.MethodGet, a, "")
+	res, err := http.DefaultClient.Do(req)
 	if err != nil {
 		return nil, err
 	}
-	defer func() {
-		err := res.Body.Close()
-		if err != nil {
-			log.Printf("error:%s", err.Error())
-		}
-	}()
-	body, _ := ioutil.ReadAll(res.Body)
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
-	}
 
-	var r ActionsResp
-	err = json.Unmarshal(body, &r.Actions)
+	parsedResponse, err := n.parseResponse(res, query.GetResponse())
```

Comment:
This is for future improvements, but I think that we can make `GetRepsonse` return a defined interface, that wrap `map[string]interface{}` and have methods that get fields to the enduser, like `GetAsString` or `GetAsFloat`...

Reply:
Created [issue](https://github.com/dhiaayachi/go-newton-co/issues/8)